examples/vision/aodnet.py

```diff
@@ -0,0 +1,362 @@
+        ),
+    )
+
+
+def random_crop(hazy_image, original_image, low_crop_size, enhanced_crop_size):
+    hazy_image_shape = tf.shape(hazy_image)[:2]
+    low_w = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[1] - low_crop_size + 1
+    )
+    low_h = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[0] - low_crop_size + 1
+    )
+    enhanced_w = low_w
+    enhanced_h = low_h
+    hazy_image_cropped = hazy_image[
+        low_h : low_h + low_crop_size, low_w : low_w + low_crop_size
+    ]
+    original_image_cropped = original_image[
+        enhanced_h : enhanced_h + enhanced_crop_size,
+        enhanced_w : enhanced_w + enhanced_crop_size,
+    ]
+    return hazy_image_cropped, original_image_cropped
+
+
+def configure_dataset(
+    dataset, image_crop_size, buffer_size, batch_size, is_dataset_train
+):
+    dataset = dataset.map(apply_scaling, num_parallel_calls=tf.data.AUTOTUNE)
+    if image_crop_size > 0:
+        dataset = dataset.map(
+            lambda hazy, original: random_crop(
+                hazy, original, image_crop_size, image_crop_size
+            ),
+            num_parallel_calls=tf.data.AUTOTUNE,
+        )
+    if is_dataset_train:
+        dataset = dataset.map(random_flip, num_parallel_calls=tf.data.AUTOTUNE)
+    dataset = dataset.shuffle(buffer_size)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.repeat(1)
+    dataset = dataset.prefetch(buffer_size=tf.data.AUTOTUNE)
+    return dataset
+
+
+def build_dataset(dataset_path, image_crop_size, buffer_size, batch_size, val_split):
+    original_images, hazy_images = get_image_file_list(dataset_path=dataset_path)
+    assert len(original_images) == len(hazy_images)
+    print("Total Number of Image Pairs:", len(original_images))
+    hazy_dataset = read_images(hazy_images)
+    original_dataset = read_images(original_images)
+    dataset = tf.data.Dataset.zip((hazy_dataset, original_dataset))
+    cardinality = tf.data.experimental.cardinality(dataset).numpy()
+    train_dataset = configure_dataset(
+        dataset=dataset.skip(int(cardinality * val_split)),
+        image_crop_size=image_crop_size,
+        buffer_size=buffer_size,
+        batch_size=batch_size,
+        is_dataset_train=True,
+    )
+    val_dataset = configure_dataset(
+        dataset=dataset.take(int(cardinality * val_split)),
+        image_crop_size=image_crop_size,
+        buffer_size=buffer_size,
+        batch_size=batch_size,
+        is_dataset_train=True,
+    )
+    return train_dataset, val_dataset
+
+
+train_dataset, val_dataset = build_dataset(
+    dataset_path="./Dehazing",
+    image_crop_size=256,
+    buffer_size=1024,
+    batch_size=16,
+    val_split=0.1,
+)
+print(train_dataset)
+print(val_dataset)
+
+"""
+## Define the AOD-Net model
+
+Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
+following diagrams from the original paper summarize its architecture:
```

Comment:
Suggestion: "its" -> "the model's" (to make sure some users don't get confused)

```suggestion
following diagrams from the original paper summarize the model's architecture:
```